analyzers/tests/SonarAnalyzer.Test/Trackers/ArgumentTrackerTest.cs

```diff
@@ -0,0 +1,1288 @@
+﻿/*
+* SonarAnalyzer for .NET
+* Copyright (C) 2015-2023 SonarSource SA
+* mailto: contact AT sonarsource DOT com
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 3 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this program; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*/
+
+using Microsoft.CodeAnalysis.Text;
+using SonarAnalyzer.Helpers.Trackers;
+using CS = Microsoft.CodeAnalysis.CSharp.Syntax;
+using VB = Microsoft.CodeAnalysis.VisualBasic.Syntax;
+
+namespace SonarAnalyzer.UnitTest.Trackers;
+
+[TestClass]
+public class ArgumentTrackerTest
+{
+    [TestMethod]
+    public void Method_SimpleArgument()
+    {
+        var snippet = """
+            System.IFormatProvider provider = null;
+            1.ToString($$provider);
+            """;
+        var (node, model) = ArgumentAndModelCS(WrapInMethodCS(snippet));
+
+        var argument = ArgumentDescriptor.MethodInvocation(KnownType.System_Int32, "ToString", "provider", 0);
+        var (result, context) = MatchArgumentCS(model, node, argument);
+        result.Should().BeTrue();
+        context.Parameter.Name.Should().Be("provider");
+        context.Parameter.Type.Name.Should().Be("IFormatProvider");
```

Comment:
These assertions are checking if the `ArgumentAndModelCS` does the right thing with the mapping. They do not validate the `ArgumentDescriptor`. I wonder if it really helps to do them every time. 

Maybe we should extract `ArgumentAndModelCS` as part of the test framework and test it independently. I do not have a strong opinion here, just that it seems a bit wrong.

Reply:
After we changed the method signatures in 39a3c1dce34c9dc186942449df9d06b4f5e93193 we do test the ArgumentContext after the matching now. I think the logic it is fine now.

Reply:
The code can be written like this:
```
        var context = ArgumentContextCS(WrapInMethodCS(snippet));
        context.Parameter.Name.Should().Be("provider");
        context.Parameter.Type.Name.Should().Be("IFormatProvider");

        var descriptor = ArgumentDescriptor.MethodInvocation(KnownType.System_Int32, "ToString", "provider", 0);
        MatchArgumentCS(context, descriptor).Should().BeTrue();
```
Which brings more clarity about what happens and when, and what the checks are verifying.

My point is that:
```
        context.Parameter.Name.Should().Be("provider");
        context.Parameter.Type.Name.Should().Be("IFormatProvider");
````
are testing `ArgumentContextCS` and only the last check `MatchArgumentCS(context, descriptor).Should().BeTrue();` is testing the `ArgumentDescriptor.MethodInvocation` method.

If you prefer to keep all the assertions, could you please move them closer to the method call? I think it would make the order clearer.

Reply:
> The code can be written like this

No, it can not:
![image](https://github.com/SonarSource/sonar-dotnet/assets/103252490/ccac5c50-bc9f-473b-9ec9-abeebfe92156)

The `new TTracker().MatchArgument(descriptor)(context);` call mutates the context. That is how the tracker works.

Reply:
That's unfortunate.